aten/src/ATen/native/mps/MPSTensorImpl.h

```diff
@@ -0,0 +1,166 @@
+    refresh_numel();
+    refresh_contiguous();
+    storage_ = at::Storage();
+  }
+
+  MPSTensorImpl(
+      at::DispatchKeySet key_set,
+      const caffe2::TypeMeta data_type,
+      c10::Device device,
+      c10::IntArrayRef sizes)
+      : TensorImpl(key_set, data_type, device) {
+    refresh_numel();
+    storage_ = at::Storage();
+  }
+
+  /**
+   * Return a TensorImpl that is a shallow-copy of this TensorImpl.
+   *
+   * For usage of `version_counter` and `allow_tensor_metadata_change`,
+   * see NOTE [ TensorImpl Shallow-Copying ].
+   */
+  c10::intrusive_ptr<TensorImpl> shallow_copy_and_detach(c10::VariableVersion&& version_counter,
+                                                         bool allow_tensor_metadata_change) const override {
+    auto impl = c10::make_intrusive<MPSTensorImpl>(key_set(), dtype(), device(), sizes());
+    copy_tensor_metadata(
+      /*src=*/this,
+      /*dest=*/impl.get(),
+      /*version_counter=*/std::move(version_counter),
+      /*allow_tensor_metadata_change=*/allow_tensor_metadata_change);
+    impl->refresh_numel();
+    impl->refresh_contiguous();
+    return impl;
+  }
+
+  /**
+   * Return a TensorImpl that is a shallow-copy of this TensorImpl.
+   *
+   * For usage of `version_counter` and `allow_tensor_metadata_change`,
+   * see NOTE [ TensorImpl Shallow-Copying ].
+   */
+  c10::intrusive_ptr<TensorImpl> shallow_copy_and_detach(const c10::VariableVersion& version_counter,
+                                                         bool allow_tensor_metadata_change) const override {
+    auto impl = c10::make_intrusive<MPSTensorImpl>(key_set(), dtype(), device(), sizes());
+    copy_tensor_metadata(
+        /*src=*/this,
+        /*dest=*/impl.get(),
+        /*version_counter=*/version_counter,
+        /*allow_tensor_metadata_change=*/allow_tensor_metadata_change);
+    impl->refresh_numel();
+    return impl;
+  }
+
+  /**
+   * Shallow-copies data from another TensorImpl into this TensorImpl.
+   *
+   * For why this function doesn't check this TensorImpl's `allow_tensor_metadata_change_`,
+   * see NOTE [ TensorImpl Shallow-Copying ].
+   */
+  void shallow_copy_from(const c10::intrusive_ptr<TensorImpl>& impl) override {
+    AT_ASSERT(has_compatible_shallow_copy_type(impl->key_set()));
+    auto mps_impl = static_cast<const MPSTensorImpl*>(impl.get());
+    copy_tensor_metadata(
+      /*src=*/mps_impl,
+      /*dest=*/this,
+      /*version_counter=*/version_counter(),
+      /*allow_tensor_metadata_change=*/allow_tensor_metadata_change());
+    refresh_numel();
+    refresh_contiguous();
+  }
+
+  const char* tensorimpl_type_name() const override;
+
+  void release_resources() override;
+
+  // storage
+  const at::Storage& storage() const override;
+
+  bool has_storage() const override {
+    return true;
+  }
+
+  bool is_contiguous(at::MemoryFormat memory_format) const override {
+    return true;
+  }
+
+ private:
+  struct Contents {
+    at::Tensor attensor_;
```

Comment:
How come an MPSTensor itself contains another Tensor, what's going on here?

Reply:
If an MPS tensor can be represented directly as the `attensor_` here, we can remove this subclass entirely, which would make me happy (CUDA tensor doesn't have a subclass)

Reply:
Agreed. We removed that 😄.